network/gopubsub/pubsub.go

```diff
@@ -0,0 +1,124 @@
+package gopubsub
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+)
+
+//PubSubChannel represents a pubsub system where subscriber can .Subscribe() and publishers can .Publish() or .Close()
+type PubSubChannel struct {
+	subscriptions []*Subscription
+	subsMutex     sync.RWMutex
+	nextId        int
+}
+
+//Subscription is created in PubSubChannel using pubSub.Ubscribe(). Subscriptors can receive using .ReceiveChannel()
+// or .Unsubscribe()
+type Subscription struct {
+	closed    bool
+	removeSub func()
+	signal    chan interface{}
+	closeOnce sync.Once
+	id        string
+}
+
+//Creates a new PubSubChannel
+func New() *PubSubChannel {
+	return &PubSubChannel{
+		subscriptions: make([]*Subscription, 0),
+	}
+}
+
+//Subscribe to a channel, each subscriptor should keep its own Subscription instance
+func (psc *PubSubChannel) Subscribe() *Subscription {
+	psc.subsMutex.Lock()
+	defer psc.subsMutex.Unlock()
+	newSubscription := newSubscription(strconv.Itoa(psc.nextId))
+	psc.nextId++
+	psc.subscriptions = append(psc.subscriptions, &newSubscription)
+	newSubscription.removeSub = func() {
+		psc.subsMutex.Lock()
+		defer psc.subsMutex.Unlock()
+
+		for i, subscription := range psc.subscriptions {
+			if subscription.signal == newSubscription.signal {
+				fmt.Println("Unsubscribing", "id", subscription.id)
+				subscription.closed = true
```

Comment:
This is the main reason I created this type and methods. The current implementation of `SubscribeToNeighbourhoodDepthChange` has a problem because you can't cancel the subscription from the subscriber side (the channel reader) because the channel should be always closed by the writer in `go` to avoid problems. 
What I do is mark the subscriptions as closed and then only close the channel when the `PubSubChannel` is stopped or a new message has been published.